src/server/application/ServiceApi.ts

```diff
@@ -0,0 +1,59 @@
+import * as express from 'express'
+import { ApolloServer } from 'apollo-server-express'
+import { gql, PubSubEngine } from 'apollo-server'
+import { Bundle, Contract, Events } from '../core'
+import { ContractController } from './ContractController'
+import { ContractRepository } from './lib/ContractRepository'
+
+export type Config = {
+  contractController: ReturnType<typeof ContractController>
+  contractRepository: ContractRepository
+  pubSubClient: PubSubEngine
+}
+
+export function ServiceApi ({ contractController, contractRepository, pubSubClient }: Config) {
+  const app = express()
+  const apolloServer = new ApolloServer({
+    typeDefs: gql`
+      type SigningRequest {
+        transaction: String!
+      }
+      type Contract {
+        engine: String!
+        contractAddress: String!
+      }
+      type Query {
+        contracts: [Contract]!
+      }
+      type Mutation {
+        loadContract(contractAddress: String!, bundleLocation: String!): Boolean
+      }
+      type Subscription {
+        transactionSigningRequest(publicKey: String!): SigningRequest
+      }
+    `,
+    resolvers: {
+      Query: {
+        async contracts () {
+          const contracts = await contractRepository.findAll()
+          return contracts.map(({ address, bundle }: { address: Contract['address'], bundle: Bundle }) => {
+            return { contractAddress: address, engine: bundle.meta.engine }
+          })
+        }
+      },
+      Mutation: {
+        loadContract (_: any, { contractAddress, bundleLocation }: { contractAddress: string, bundleLocation: string }) {
+          return contractController.load(contractAddress, bundleLocation)
+        }
+      },
+      Subscription: {
+        transactionSigningRequest: {
+          subscribe: (_: any, { publicKey }: { publicKey: string }) => pubSubClient.asyncIterator(`${Events.SIGNATURE_REQUIRED}.${publicKey}`)
+        }
+      }
+    },
+    introspection: true
+  })
+  apolloServer.applyMiddleware({ app })
```

Comment:
What's the purpose of this?

I see you are returning the express instance. Is this only to adhere to the http.Server type for shutdown?

If so, according to the docs (https://www.apollographql.com/docs/apollo-server/api/apollo-server) new ApolloServer().listen() returns a promise, that when resolved is an object with a server key that is of the http.Server type

Reply:
Just added annotations to explain this